keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+no startup probe defined, postStart completed).
+
+These containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+Those containers will also be restarted even when the Pod restartPolicy is Never
+or OnFailure. Furthermore, sidecar containers will be restarted even during Pod
+termination.
+
+In order to minimize OOM kills of sidecar containers, the OOM adjustment for
+these containers will match or exceed the OOM score adjustment of regular
+containers in the Pod. This intent to solve the issue
+https://github.com/kubernetes/kubernetes/issues/111356 
+
+As part of this KEP we also will be enabling for sidecar containers: PreStart
+and PostStop lifecycle handlers for sidecar containers All probes (startup,
+readiness, liveness) Readiness probes of sidecars will be used to determine the
+whole Pod readiness, but only after the main containers initialization started.
+Readiness probes will not start running before the Pod initialization stage
+completed 
+
+```yaml
+kind: Pod
+spec:
+  initContainers:
+  - name: vault-agent
+    image: hashicorp/vault:1.12.1
+    restartPolicy: OnFailure
+  - name: istio-proxy
+    image: istio/proxyv2:1.16.0
+    args: ["proxy", "sidecar"]
+    restartPolicy: Always
+  containers:
+  ...
+```
+
+### Future use of restartPolicy field
+
+In the Alpha version, no other values beyond Always will be acceptable. Going
+forward the following semantic may be supported for other values of
+restartPolicy field of Init containers. The restartPolicy may be set to one of:
+`Never`, `OnFailure` and `Always`.  The meaning of these values is similar to
+the Pod 'restartPolicy' flag:
+- `Never` indicates that if the container returns a non-zero exit code, that pod
+  initialization has failed. This value can be supported for Pods with Never and
+  OnFailure restartPolicy indicating the critical Initialization step.
+  Supporting this value on Pods with the the restartPolicy “Always” is not
+  planned as it changes the semantic of this policy and may affect how
+  deployments control the # of Pods in this deployment.
+- `OnFailure` indicates the init container will be restarted until it completes
+  successfully. This value may be used for Pods with the restartPolicy “Never” ,
+  indicating that the Initialization step is flaky and should be retried.
+- `Always` indicates that the container will be considered initialized the
+  moment the startup probe indicates it is started and will be restarted if it
+  exits for any reason until the pod terminates.
+- (If the initContainer's restartPolicy is unset, the pod's restartPolicy is
+  used to determine the Init container restart policy. For Always, OnFailure is
+  used, for OnFailure - OnFailure, for Never - Never).
+
+Since Init container and regular containers share the schema, restartPolicy
+field will be available for regular containers as well. From scenarios
```

Comment:
This section is a bit confusing.  I think we need to clarify it a bit.

1. Are you intending for regular containers to have restartPolicy (as in, we have a future use case for it)?
  * If we aren't, sharing the schema is problematic and it might be better to do what ephemeral containers did and split the schemas
  * If we are, from a design perspective this sounds very confusing and I'd have a few suggestions
2. Are you intending to default the "inferred" container restart policy value based on pod policy into each container?

Reply:
No matter what as a result of this change we need to have better API doc on init container fields, so it's probably time to split them.

Reply:
IMO we should not be sharing types BUT we also make it very difficult NOT to do that.

Reply:
Initiated the discussion: https://github.com/kubernetes/kubernetes/issues/115362